### PR TITLE
PIM-6676: add documentation explaining frontend architecture

### PR DESCRIPTION
**Pull request by juliensnz**

### CODE REVIEW

File: design_pim/overview.rst [hunk 1]
```diff
@@ -0,0 +1,322 @@
+How to customize any frontend part of the application
+-----------------------------------------------------
+
+Introduction
+===================
+
+Akeneo PIM uses requeirejs and backbone to manage modules and views in our application. On top of those two layers, we added our own layer: the form extension system. In this documentation, we will discover the basics of those three layers to be able to understand Akeneo PIM stack and customize it.
```
Comment (nidup) [inline design_pim/overview.rst]: requirejs

File: design_pim/overview.rst [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+You can also decide to symlink those assets instead of copying them:
+
+    bin/console assets:install --symlink web
+
+After that the last step is to build the bundle.js file
+
+    npm run webpack
+
+This command will compile and minify all the pim files into one unique file and dump it in the public folder. You can also use the watch command which will recompile this file each time you modify a registered module.
```
Comment (tamarasaurus) [inline design_pim/overview.rst]: It dumps a few bundles, one for library modules, vendor, webpack util stuff and then the main app code

File: design_pim/overview.rst [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+===============================
+
+Now that we can create a requirejs module and a backbone view, we would like to actually customize the PIM to add our own business logic on top of it.
+
+A little bit of history
++++++++++++++++++++++++
+
+As we are an open source company, our product can used and customized for a lot of different reasons and by a lot of different people (integrators, clients, contributors, technological partners). With this in mind, it was impossible to base our architecture on overrides to customize it.
+
+We decided to create a tree based architecture where each form of the application would be a tree of extensions. If you want to change a part of a page or add something to it, you need to create a requirejs module and register it in the tree of form extensions.
```
Comment (mickaelandrieu) [inline design_pim/overview.rst]: If I want to hide/change/replace a form - lets say in Product Edit form, very common use case - how can we do?

Comment (a2xchip) [inline design_pim/overview.rst]: Just create new form extension copy/pasting existent and changing it... Same id plus higher bundle hierarchy ;-)

Comment (mickaelandrieu) [inline design_pim/overview.rst]: How can I know the id for a field created by a reference data, this is probably from a template and I don't want to update all forms, only "this one": is there a common naming rule? Let's say if my reference data is named "toto" and linked to the product inside the tab "tata", the id of the related form is .. ?

Comment (juliensnz) [inline design_pim/overview.rst]: It's not linked to the reference data, it's linked to the fieldProvider. To override a field, here is the documentation: https://docs.akeneo.com/1.7/cookbook/ui_customization/create_a_new_field_type.html#create-a-field-provider

With this documentation, you can do what you are asking for :)

File: design_pim/overview.rst [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+.. code-block:: yaml
+
+    # src/Acme/Bundle/CustomBundle/Resources/config/requirejs.yml
+    config:
+        paths:
+            my_locale_logger: acmecustom/js/locale_logger
+
+
+We just declared a new module named ``my_locale_logger`` located in ``web/bundles/acmecustom/js/locale_logger.js``.
+Note that we don't need to specify the file extension and it is located in the public web folder (moved during the ``bin/console assets:install`` command execution). Also, ``my_locale_logger`` name is a free name and you can replace it with whatever you want.
```
Comment (phaseinducer) [inline design_pim/overview.rst]: "arbitrary" instead of "free" ?

File: design_pim/overview.rst [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+Summary
++++++++
+
+We just created our first module and made it available for other part of the application. You can now use it by requirering ``my_locale_logger``.
+
+
+Create our first backbone view
+==============================
+
+Know that we know how to create a requirejs module, let's create our first Backbone view.
```
Comment (phaseinducer) [inline design_pim/overview.rst]: know -> now

File: design_pim/overview.rst [hunk 6]
```diff
@@ -0,0 +1,327 @@
+How to customize any frontend part of the application
+-----------------------------------------------------
+
+Introduction
+===================
+
+Akeneo PIM uses requirejs and backbone to manage modules and views in our application. On top of those two layers, we added our own layer: the form extension system. In this documentation, we will discover the basics of those three layers to be able to understand Akeneo PIM stack and customize it.
```
Comment (phaseinducer) [inline design_pim/overview.rst]: "RequireJS"
"Backbone"
(with capitals) or make it lowercase everywhere else for consistency.

File: design_pim/overview.rst [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+        ['backbone'],
+        function (Backbone) {
+            return Backbone.View.extend({
+
+            });
+        }
+    );
+
+
+This is the simplest view we can create using backbone. It will result into an empty div if we add it to the DOM.
```
Comment (phaseinducer) [inline design_pim/overview.rst]: The first sentence is weird, what do you mean ?

File: design_pim/overview.rst [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+
+Useful methods
+++++++++++++++
+
+Here are a list of method that you can override or call in your extension that should make your life easier.
+
+Managing the model
+******************
+
+Each form have an internal model representing the current object we are modifying. Here is the way to access it and modify it
```
Comment (phaseinducer) [inline design_pim/overview.rst]: You should say that `getFormData()` doesn't return the model of the view itself but the root's one.

File: design_pim/overview.rst [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+        }
+    });
+
+
+As you can see, we can get the model by calling ``this.getFormData()`` from any extension and update the model with ``this.setData()``. Not that calling setData will trigger the event ``pim_enrich:form:entity:(pre|post)_update`` on the root view. You can set the option 'silent' to true on the to avoid triggering it (``this.setData(data, {silent: true})``).
+
+Configure your extension
+************************
+
+Some times, you want to perform actions before the first render (fetch informations, do heavy computation, etc). The configure method is perfect for this need
```
Comment (phaseinducer) [inline design_pim/overview.rst]: "Some times" -> "Sometimes"

Comment (phaseinducer) [reply design_pim/overview.rst]: End your sentence by a period or a colon please.

File: design_pim/overview.rst (hunk 7 above)
Comment (juliensnz) [reply design_pim/overview.rst]: I don't know: this is the simplest view we can create using backbone framework. I don't understand what you don't understand :/

File: design_pim/overview.rst [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+            return {
+                logLocale: function () {
+                    console.log(UserContext.get('uiLocale'));
+                }
+            }
+        }
+    )
+
+
+The ``define`` method takes two arguments: the first one is an array of the dependencies we want to use (``pim/user-context`` in this case) and a callback that will be called every time our module will be requested. This callback will receive as argument the list dependencies we asked for. Note that you can name the arguments of the callback as you want.
```
Comment (pierallard) [inline design_pim/overview.rst]: Please add where the pim/user-context is defined (the .yml and the .js files for real)

File: design_pim/overview.rst [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+
+.. code-block:: bash
+
+    npm run webpack-watch
+
+
+Summary
++++++++
+
+We just created our first module and made it available for other part of the application. You can now use it by requirering ``my_locale_logger``.
```
Comment (pierallard) [inline design_pim/overview.rst]: It miss "how to use it"... In fact, yes, we created a module, but we didn't see it in action.

File: design_pim/overview.rst [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+        my_form_extension:
+            module: my_custom_view             # Your RequireJS module name
+            parent: pim-product-edit-form-meta # The parent of your extension (the meta section of the product edit form in this case)
+
+
+A fiew words about this small configuration:
+
+- the key ``my_form_extension`` should be a unique key to represent your form extension.
+- you can declare multiple form extensions with the same RequireJS module.
+- to be registered your configuration file should be named ``form_extensions.yml`` or putted in a ``form_extensions`` folder in your bundle ``Resources/config`` folder.
```
Comment (pierallard) [inline design_pim/overview.rst]: have to be named

File: design_pim/overview.rst [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+This configuration is the minimal example. Here are the other parameters that you can use:
+
+.. code-block:: yaml
+
+    # src/Acme/Bundle/CustomBundle/Resources/config/form_extensions.yml
+    extensions:
+        my_form_extension:
+            module: my_custom_view
+            parent: pim-product-edit-form-meta
+            targetZone: header       # Where to put the extension in the parent DOM
```
Comment (pierallard) [inline design_pim/overview.rst]: [optional]

File: design_pim/overview.rst [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+            position: 90             # Its position among the other extensions
```
Comment (pierallard) [inline design_pim/overview.rst]: [optional] .. etc

File: design_pim/overview.rst (hunk 10 above)
Comment (juliensnz) [reply design_pim/overview.rst]: As this is just an example and not relevant, I'm not sure we should precise it. This module can be registered anywhere

File: design_pim/overview.rst (hunk 11 above)
Comment (juliensnz) [reply design_pim/overview.rst]: Yeah I agree about that, but as we cannot use it until line 219, we are kind of stuck. I will add an explanation about that :)

File: design_pim/overview.rst (hunk 14 above)
Comment (juliensnz) [reply design_pim/overview.rst]: In fact they are all optional and I explain just above that this is a minimal example. I don't know if I need to add it here again.

File: design_pim/overview.rst (hunk 8 above)
Comment (juliensnz) [reply design_pim/overview.rst]: DOne :)

File: design_pim/overview.rst [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+    rm -rf ./var/cache/*
+
+
+dump the assets in the web folder
+
+.. code-block:: bash
+
+    bin/console assets:install web
+
```
Comment (tamarasaurus) [inline design_pim/overview.rst]: I think we need to put pim:installer:dump-require-paths after this

Comment (juliensnz) [reply design_pim/overview.rst]: I never ran this command so maybe it's not mandatory?

Comment (tamarasaurus) [reply design_pim/overview.rst]: Oh that's right it gets run in the pim install command

File: design_pim/overview.rst [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+Create our first form extension
+===============================
+
+Now that we can create a RequireJS module and a Backbone view, we would like to actually customize the PIM to add our own business logic on top of it.
+
+A little bit of history
++++++++++++++++++++++++
+
+As we are an open source company, our product can used and customized for a lot of different reasons and by a lot of different people (integrators, clients, contributors, technological partners). With this in mind, it was impossible to base our architecture on overrides to customize it.
```
Comment (gplanchat) [inline design_pim/overview.rst]: Fix typo : As we are an open source company, our product can **be** used and customized for a lot of different reasons and by a lot of different people (integrators, clients, contributors, technological partners). With this in mind, it was impossible to base our architecture on overrides to customize it.

**Comment by gplanchat** [comment]
One question I have is : How to create a new form from scratch ?

eg. for reference data or custom entities (in general, not limited to `PimCustomEntityBundle`)

**Pull request merged**